tests/12login/02cas.pl

```diff
@@ -0,0 +1,143 @@
+      my ( $http, $homeserver_info ) = @_;
+
+      my $HS_URI = $homeserver_info->client_location;
+
+      # the redirectUrl we send to /login/cas/redirect, which is where we
+      # hope to get redirected back to
+      my $REDIRECT_URL = "https://client?p=http%3A%2F%2Fserver";
+
+      # the ticket our mocked-up CAS server "generates"
+      my $CAS_TICKET = "goldenticket";
+
+      # step 1: client sends request to /login/cas/redirect
+      # step 2: synapse should redirect to the cas server.
+      Future->needs_all(
+         wait_for_cas_request("/cas"),
+         $http->do_request(
+            method => "GET",
+            uri    => "/api/v1/login/cas/redirect",
+            params => {
+               redirectUrl => $REDIRECT_URL,
+            },
+         ),
+      )->then( sub {
+         my ( $cas_request, $cas_response ) = @_;
+         log_if_fail( "Initial CAS request query:",
+                      $cas_request->query_string );
+
+         assert_eq( $cas_request->method, "GET", "CAS request method" );
+
+         my $service = $cas_request->query_param( "service" );
+
+         # step 3: client sends credentials to CAS server
+         # step 4: CAS redirects, with a ticket number, back to synapse.
+         #
+         # For this test, we skip this bit, as it's nothing to do with synapse,
+         # really.
+         #
+         # The URI that CAS redirects back to is the value of the 'service'
+         # param we gave it, with an additional "ticket" query-param:
+         my $login_uri = $service . "&ticket=$CAS_TICKET";
+
+         # step 5: synapse receives ticket number from client, and makes a
+         # request to CAS to validate the ticket.
+         # step 6: synapse sends a redirect back to the browser, with a
+         # 'loginToken' parameter
+         Future->needs_all(
+            wait_for_cas_request(
+               "/cas/proxyValidate",
+               response => $CAS_SUCCESS,
+            ),
+            $http->do_request_json(
+               method   => "GET",
+               full_uri => $login_uri,
+               max_redirects => 0, # don't follow the redirect
+              )->followed_by( \&main::expect_http_302 ),
+         );
+      })->then( sub {
+         my ( $cas_validate_request, $ticket_response ) = @_;
+
+         assert_eq( $cas_validate_request->method, "GET",
+                    "/cas/proxyValidate request method" );
+         assert_eq( $cas_validate_request->query_param( "ticket" ),
+                    $CAS_TICKET,
+                    "Ticket supplied to /cas/proxyValidate" );
+         assert_eq( $cas_validate_request->query_param( "service" ),
+                    $HS_URI,
+                    "Service supplied to /cas/proxyValidate" );
+
+         my $redirect_uri = URI->new($ticket_response->header( "Location" ));
+         log_if_fail("Redirect from /login/cas/ticket",
+                     $redirect_uri);
+         assert_eq( index($redirect_uri, "https://client?"), 0,
```

Comment:
A valid way to test, but in other places around sytest we've tended towards a regexp match rather than `index()`; so maybe something more like

    assert_ok( $redirect_uri =~ m/^https://client\?/,